Still report / read loaded virtualBuffer after too early focus event in Chromium 100

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft. See https://github.com/nvaccess/nvda/wiki/Contributing
-->

### Link to issue number:
Fixes #13571 
Fixes #13570 

### Summary of the issue:
Google Chrome 100, and presumably MS Edge 100 which hasn't been released at the time of this issue creation, has broken some aspects of NVDA's browsing capabilities and features, resulting in the suppression of the announcement of the page title upon load, and malfunctioning of the Automatic SayAll on Page Load feature. This also tampers with the capabilities of some add-ons -- namely BrowserNav.
It seems that in Chromium 100 there is a definite change to the events we receive when a page loads.
It looks like the following happens:
1. the user causes a page to load by pressing enter in the location bar or activating a link
2. A new document accessible is created with no name, no value, and states of focused, focusable and readonly.
3. a focus event is fired on this document.
3.1. NVDA receives the focus event, and tries to load a virtualBuffer for the document, but it is blank (no content) and NVDA cannot get its URL (documentConstantIdentifier) as the document's accValue is empty still.
4. Content starts to load, and the name, value, and descendants are added.
5. A documentLoadComplete event is fired on the document.
6.1. Currently NVDA does not pay attention to this event, as it expects that the first or another focus event would happen when most if not all of the document was loaded.
Previously, focus was not fired on the document until the name, value, and at least some of the descendants were loaded. And therefore, NVDA was able to load a virtualBuffer with at least some content.

### Description of how this pull request fixes the issue:
Added an implementation of event_documentLoadComplete to the gecko_ia2 virtualBuffer class, which performs the initial reporting / sayAll on virtualBuffer load, if this has not yet occurred.
Chromium fires a documentLoadComplete event once the page is fully loaded. Thus by then all the content will be rendered in our virtualBuffer.
Also, BrowseModeDocument's documentConstantIdentifier is no longer forcefully cached on construction, rather each of the implementations of this property, cache themselves the first time they fetch a non-empty value. This works around the fact that Chromium 100 has an empty accValue when the document is first created.

### Testing strategy:
* Opened google Chrome 100
* In a new tab, visited: https://www.nvaccess.org/
* Ensured that once the page loaded, NVDA automatically announced the page title and started reading the page.
* Tested several other pages in Mozilla Firefox to ensure that browsers with existing page load logic still function correctly.

### Known issues with pull request:
None known.

### Change log entries:

Bug fixes:
```
- NVDA will no longer fail to report the page title and start automatically reading, when a page loads in Google chrome 100. (#13571)
```


### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

michaelDCurran committed:
base virtualBuffer: add an event_documentLoadComplete method which handles the case where initial reporting of the virtualBuffer has not occured yet due to focus event being fired too early, such as in Chromium 100.

feerrenrut reviewed (dismissed):
Looks good. You might consider adding comments to the `_get_documentConstantIdentifier` to point out that ID is being cached intentionally only if it is valid and why.

I suspect this change will resolve the occasional system test failures for chrome also.

michaelDCurran commented:
@feerrenrut  I've added a comment to gecko_ia2 virtualBuffer's documentConstantIdentifier property now. However, I have chosen to remove the if condition from all other implementations of documentConstantIdentifier, caching ID as is without checking, as this mirrors the original behavior.

feerrenrut commented:
I note that the tests are now passing. The older appveyor comments are now hidden.

michaelDCurran committed:
Gecko_ia2 vbufBackend: similar to with focus and alert events, documentLoadComplete event also now forces any existing nodes marked as invalid to be updated straight away.
Necessary as NVDA may try to start reading the page on a documentLoadComplete event, and therefore the buffer must be fully up to date.

michaelDCurran committed:
BrowseModeDocumentTreeInterceptor:
* Mark the documentConstantIdentifier property for caching for the current core cycle.

* Rather than forcefully overriding the documentConstantIdentifier property with its own value on class initialization, cache its value on a new _lastCachedDocumentConstantIdentifier instance variable. Also update this variable each time the caret moves, as the identifier (URL) may have changed dynamically (such as with a single page app).
* _getInitialCaretPos and terminate now use _lastCachedDocumentConstantIdentifier rather than documentConstantIdentifier. This means that they have the most up to date value, but also won't fail if the document is already dead.

michaelDCurran committed:
Fix linting

michaelDCurran committed:
Merge branch 'beta' into i13571

michaelDCurran commented:
@feerrenrut  I've added a few more changes.
Firstly, the gecko_ia2 vbufBackend now forces any pending updates to the buffer to run straight away if a documentLoadComplete event is received. This is similar to focus and alert events. This is necessary as NVDA expects that the buffer is fully up to date on these events, as it may try to start a sayAll or restore the caret position etc. Previously in Chrome when going back to a previously visited page NVDA would just say blank as the buffer was not yet fully updated.
I have also re-thought the changes to documentConstantIdentifier, both removing the need to needlessly change all the implementations, but also solving the issue of NVDA not being able to restore the virtual caret  position when going back to a page in a SPA (single page app) such as Github pull request queries. 
In short: Originally we would forcefully override the documentConstantIdentifier property with its own value on class initialization, essentially caching it for ever, so that on terminate we could save off the last caret position in the appModule keyed by the documentConstantIdentifier for later restoring the virtual caret if the user went back to that pgage. that value had to be cached as in terminate the document accessible would already be dead.
However, this never took into account two things:
1. The document's URL could change (E.g. due to an SPA) and
2. 2. the document may not have a valid URL until it is fully loaded (such as with Chromium 100).

As the change is in the base BrowseModeDocumentTreeInterceptor, all browsers will be able to take advantage of these fixes, even though we've only seen the second issue in Chromium 100 so far.

feerrenrut commented on nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp:
```diff
@@ -1208,8 +1209,12 @@ void CALLBACK GeckoVBufBackend_t::renderThread_winEventProcHook(HWINEVENTHOOK ho
+		if(
+			eventID == EVENT_OBJECT_FOCUS
+			|| eventID == IA2_EVENT_DOCUMENT_LOAD_COMPLETE
+			||eventID==EVENT_SYSTEM_ALERT
```
nit: space after boolean OR 
```suggestion
			|| eventID==EVENT_SYSTEM_ALERT

```

feerrenrut commented on source/browseMode.py:
```diff
@@ -1778,7 +1795,10 @@ def doResult(result):
 		return doResult(False)
 
-	def _get_documentConstantIdentifier(self):
+	documentConstantIdentifier: Optional[str]
```
Auto property docstring
```suggestion
	documentConstantIdentifier: Optional[str]
	""" Typing information for auto-property: _get_documentConstantIdentifier"""
```

feerrenrut commented on source/browseMode.py:
```diff
@@ -1778,7 +1795,10 @@ def doResult(result):
 
-	def _get_documentConstantIdentifier(self):
+	documentConstantIdentifier: Optional[str]
+	# Mark documentConstantIdentifier property   for caching during the current core cycle
+	_cache_documentConstantIdentifer = True
```
I think doc strings like this are supposed to be a string after the property. I'd also refer to "auto-property" for consistency with the typing doc strings.
```suggestion
	_cache_documentConstantIdentifer = True
	""" Cache auto-property: documentConstantIdentifier during a core cycle. """
```

feerrenrut commented on source/browseMode.py:
```diff
@@ -1778,7 +1795,10 @@ def doResult(result):
+	documentConstantIdentifier: Optional[str]
+	# Mark documentConstantIdentifier property   for caching during the current core cycle
+	_cache_documentConstantIdentifer = True
+	def _get_documentConstantIdentifier(self) -> Optional[str]:
 		"""Get the constant identifier for this document.
```
Given this ID can change, the 'constant' in the name is a bit deceptive. Can we at least point this out in the doc-string?

feerrenrut reviewed (commented):
So this will re-cache the doc ID on gain focus, set selection. Now, internally `_lastCachedDocumentConstantIdentifier` will be used.

- What about code that still uses the `documentConstantIdentifier` property? Just checked, it seems usage is limited to overrides and `browseMode.py` for core code. What about addons?
- Is there a situation where the cache should be invalidated, I.E. it has previously been set, but `documentConstantIdentifier` now returns None?

michaelDCurran committed:
Apply suggestions from code review

Co-authored-by: Reef Turner <feerrenrut@users.noreply.github.com>

michaelDCurran committed:
Update docString

michaelDCurran committed:
Update what's new

XLTechie commented:
@michaelDCurran In browseMode.py, line 1804:
```
	_cache_documentConstantIdentifer = True
```

Shouldn't that be spelled "Identifier"?

I tried to comment this on the PR itself, but it was already closed so couldn't.

michaelDCurran commented:
@XLTechie   Thanks for noticing that. Fixed in commit 66b5f24fa936
